### PR TITLE
Extend node e2e test suite with containerized Kubelet

### PR DESCRIPTION
**Pull request by ingvagabund**
How to run it?
```sh
make test-e2e-node \
    TEST_ARGS='--kubelet-containerized=true --hyperkube-image=hyperkube-amd64:1.9 --kubelet-flags="<FLAGS>"' \
    FOCUS="Conformance"
```

### CODE REVIEW

**Comment by ingvagabund** [comment]
How relevant is it to implement the non-systemd variant?

File: test/e2e_node/services/kubelet.go [hunk 1]
```diff
@@ -125,7 +133,34 @@ func (e *E2EServices) startKubelet() (*server, error) {
...
+		if kubeletContainerized {
+			cmdArgs = append(cmdArgs, systemdRun, "--unit="+unitName, "--slice=runtime.slice", "--remain-after-exit",
+				"/usr/bin/docker", "run", "--name=kubelet",
+				"--rm", "--privileged", "--net=host", "--pid=host",
+				"-e HOST=/rootfs", "-e HOST_ETC=/host-etc",
+				"-v", "/etc/localtime:/etc/localtime:ro",
+				"-v", "/etc/machine-id:/etc/machine-id:ro",
+				"-v", "/etc/systemd/system:/host-etc/systemd/system",
+				"-v", filepath.Dir(kubeconfigPath)+":/etc/kubernetes",
+				"-v", "/lib/modules:/lib/modules",
```
Comment (sjenning) [inline test/e2e_node/services/kubelet.go]: loading kernel modules?

File: test/e2e_node/services/kubelet.go [hunk 2]
```diff
@@ -125,7 +133,34 @@ func (e *E2EServices) startKubelet() (*server, error) {
...
 		unitName := fmt.Sprintf("kubelet-%d.service", rand.Int31())
-		cmdArgs = append(cmdArgs, systemdRun, "--unit="+unitName, "--slice=runtime.slice", "--remain-after-exit", builder.GetKubeletServerBin())
+		if kubeletContainerized {
+			cmdArgs = append(cmdArgs, systemdRun, "--unit="+unitName, "--slice=runtime.slice", "--remain-after-exit",
+				"/usr/bin/docker", "run", "--name=kubelet",
+				"--rm", "--privileged", "--net=host", "--pid=host",
+				"-e HOST=/rootfs", "-e HOST_ETC=/host-etc",
+				"-v", "/etc/localtime:/etc/localtime:ro",
+				"-v", "/etc/machine-id:/etc/machine-id:ro",
+				"-v", "/etc/systemd/system:/host-etc/systemd/system",
```
Comment (sjenning) [inline test/e2e_node/services/kubelet.go]: what is this needed for?

**Comment by sjenning** [comment]
> How relevant is it to implement the non-systemd variant?

What specifically makes this systemd only?

File: test/e2e_node/services/kubelet.go (hunk 1 above)
Comment (ingvagabund) [reply test/e2e_node/services/kubelet.go]: I got inspired by the openshift-node service file. If not needed, I can remove it.

File: test/e2e_node/services/kubelet.go (hunk 2 above)
Comment (ingvagabund) [reply test/e2e_node/services/kubelet.go]: I got inspired by the openshift-node service file. If not needed, I can remove it.

**Comment by ingvagabund** [comment]
> What specifically makes this systemd only?

If the systemd-run binary is present, the systemd service is created. Otherwise, a binary on a background is run. So it is needed by the non-systemd OSes.

File: test/e2e_node/services/kubelet.go (hunk 1 above)
Comment (sjenning) [reply test/e2e_node/services/kubelet.go]: I think we can remove these two (at least).

**Comment by sjenning** [comment]
> If the systemd-run binary is present, the systemd service is created. Otherwise, a binary on a background is run. So it is needed by the non-systemd OSes.

Ah I see.  So you are just adding support for the systemd path in this PR atm.

How much trouble would it be to implement the non-systemd path as well?  This might help our case for upstream containerized kubelet testing.

**Comment by ingvagabund** [comment]
@dchen1107 PTAL

**Comment by derekwaynecarr** [comment]
this looks fine to me.

/lgtm
/approve

File: test/e2e_node/services/kubelet.go [hunk 3]
```diff
@@ -93,6 +97,10 @@ const (
 // startKubelet starts the Kubelet in a separate process or returns an error
 // if the Kubelet fails to start.
 func (e *E2EServices) startKubelet() (*server, error) {
+	if kubeletContainerized && hyperkubeImage == "" {
+		return nil, fmt.Errorf("the --hyperkube-image option must be set")
```
Comment (ScorpioCPH) [inline test/e2e_node/services/kubelet.go]: nit: the --hyperkube-image option must be set if we run kubelet in a docker container

**Review by ScorpioCPH — COMMENTED** [review]
lgtm with minor nit comments :)

File: test/e2e_node/services/kubelet.go (hunk 3 above)
Comment (ingvagabund) [reply test/e2e_node/services/kubelet.go]: The error message pops up only when the ``--containerized`` flag is set. Later, the ``--hyperkube-image`` can point to a container of non-docker container runtime. There is always a space for ambiguity. However, the ``hyperkube`` + ``image`` connection is currently strongly voting for containerized run, the confusion will be minimal. But yeah, updating the message would help to make the statement more clear :).

Comment (ScorpioCPH) [reply test/e2e_node/services/kubelet.go]: SGTM.

**Pull request merged**